Thread Wrappers

just-max opened the pull request:
Extends the concept of `malloc` wrappers to thread creation. Threads receive an ID based on the CFG node of the wrapper function's call, instead of the `pthread_create` call, which in the case of a wrapper is always the same node.

Closes #664.

michael-schwarz commented on src/analyses/threadId.ml:
```diff
@@ -104,13 +104,19 @@ struct
+
   let threadenter ctx lval f args =
-    let+ tid = create_tid ctx.local ctx.prev_node f in
+    (* x *)
```
why the `(* x *)`?

just-max replied:
The PR is still in a draft state, needs tests and a cleanup still!

michael-schwarz commented on src/analyses/threadId.ml:
```diff
@@ -104,13 +104,19 @@ struct
   let threadspawn ctx lval f args fctx =
     let (current, td) = ctx.local in
-    (current, Thread.threadspawn td ctx.prev_node f)
+    (* x *)
```
Same here

michael-schwarz commented on src/domains/queries.ml:
```diff
@@ -117,6 +126,7 @@ type _ t =
   | IterSysVars: VarQuery.t * Obj.t VarQuery.f -> Unit.t t (** [iter_vars] for [Constraints.FromSpec]. [Obj.t] represents [Spec.V.t]. *)
   | MayAccessed: AccessDomain.EventSet.t t
   | MayBeTainted: LS.t t
+  (* |  *)
```
Why `(* | *)`?

michael-schwarz commented on src/analyses/wrapperFunctionAnalysis.ml:
```diff
@@ -0,0 +1,208 @@
+        remove unique_call counter |> add unique_call (count + 1)
+  end
+
+  module ThreadNode = struct
+    include Printable.Prod3 (ThreadIdDomain.ThreadLifted) (Node) (Chain)
+
+    (* Description that gets appended to the varinfo-name in user output. *)
+    let describe_varinfo (v: varinfo) (t, node, c) =
+      let loc = UpdateCil.getLoc node in
+      CilType.Location.show loc
+
+    let name_varinfo (t, node, c) =
+      Format.asprintf (* TODO *) "(alloc@sid:%s@tid:%s(#%s))" (Node.show_id node) (ThreadLifted.show t) (Chain.show c)
+
+  end
```
I don't quite get what this is for yet.

michael-schwarz commented:
We now have an issue with `Duplicate test ID 66/01` here. The reason is that master got a group 66 in the meanwhile, so we should change the group number to the next free one.

michael-schwarz commented on src/domains/queries.ml:
```diff
@@ -13,6 +13,15 @@ module TS = SetDomain.ToppedSet (CilType.Typ) (struct let topname = "All" end)
 module VS = SetDomain.ToppedSet (CilType.Varinfo) (struct let topname = "All" end)
 
+
+(* TODO: where to put this *)
```
I think here is fine!

michael-schwarz commented on src/util/debug.ml:
```diff
@@ -0,0 +1,6 @@
+open GoblintCil
+
+let f = Printf.sprintf
+let pf fmt = Printf.ksprintf print_endline fmt
+let df fmt = Pretty.gprintf (Pretty.sprint ~width:max_int) fmt
+let dpf fmt = Pretty.gprintf (fun doc -> print_endline @@ Pretty.sprint ~width:max_int doc) fmt
```
Instead of introducing this extra mechanism, it might be nice to rely on Tracing which we already use in a lot of places throughout Goblint.

just-max replied:
Oops, this is still temporary debugging code.

michael-schwarz commented on src/analyses/wrapperFunctionAnalysis.ml:
```diff
@@ -0,0 +1,257 @@
+  let dbg = WrapperArgs.is_wrapped (LibraryDesc.ThreadCreate { thread = Cil.integer 1; start_routine = Cil.integer 1; arg = Cil.integer 1; })
+  let st name ctx = if dbg then dpf"----------------------------------\n[%s] prev_node=%a => node=%a" name Node.pretty ctx.prev_node Node.pretty ctx.node
+  
+  (* TODO:
+    Does it matter if this is node or prev_node? [malloc] analysis used ctx.node and seemed to care.
+    Thread ID analysis is using ctx.prev_node (which makes more sense, since that's where the thread_create edge is,
+    and would keep two wrapper calls apart if they are e.g. both on edges leading into a join point)
+    https://github.com/goblint/analyzer/commit/77c0423640c50bb82e4290bcc97f33d4082715d0
+      *)
```
Intuitively it makes more sense to use `prev_node`. Do things break if one uses that everywhere?

just-max replied:
The malloc analysis breaks :) You changed it from prev_node in the linked commit, but I can't figure out why the change was made.

just-max replied:
We determined that the malloc analysis doesn't break (I still had an old test run in mind), so I'll change this to always use prev_node.

michael-schwarz commented on src/analyses/wrapperFunctionAnalysis.ml:
```diff
@@ -0,0 +1,257 @@
+        match wrapper_node with
+        (* if an interesting callee is called by an interesting caller, then we remember the caller context *)
+        | `Lifted _ -> wrapper_node
+          (* if dbg then dpf"interesting caller, keep caller context"; *)
```
Please do not add anymore commented out code than we already have :wink:

michael-schwarz commented on src/analyses/wrapperFunctionAnalysis.ml:
```diff
@@ -0,0 +1,257 @@
+          (* if dbg then dpf"interesting caller, keep caller context"; *)
+        (* if an interesting callee is called by an uninteresting caller, then we remember the callee context *)
+        | _         -> `Lifted (node_for_ctx ctx)
+          (* if dbg then dpf"uninteresting caller, keep callee context"; *)
```
Same here

michael-schwarz commented on src/analyses/wrapperFunctionAnalysis.ml:
```diff
@@ -0,0 +1,257 @@
+
+  let add_unique_call ctx =
+    let counter, wrapper_node = ctx.local in
+    (* TODO: previously, unique count isn't by wrapper node (e.g. my_malloc_wrapper), but by wrapped node (e.g. malloc). Why, and is it safe to change? *)
```
This is probably incidental and due to the student who made the implementation of the unique counter. Should be safe to change!

just-max commented:
A last note and implementation change I've made: `threadcreate` and `threadspawn` see different contexts: the former before `special` updates it (and increments the unique-counter), the latter after. The fix up until now didn't properly handle reaching `Top`. I've changed it so that the wrapper-function domain stores a pair of counter values `(previous, current)` for each call site; in `threadcreate` we get the current value, in `threadspawn` the previous value. Note that just adding/subtracting 1 to work around this isn't enough, since when the current counter is `Top`, the previous counter could either be `Top - 1` or `Top` with no way of knowing which.

This should be ready now. I've tested this PR with `gobopt='--set ana.activated[+] threadCreateWrapper --set ana.thread.unique_thread_id_count N' scripts/update_suite.rb` for N=0,1,8 with no problems (apart from the obvious failures from two of the new tests that expect N>=1). The new tests are in regression/71... instead of regression/70... anticipating that #850 will be merged first :) What's left: once there are some benchmarks for this, we could consider having it on be default.

Note that, like the malloc wrapper analysis, a given N actually allows distinguishing N+1 calls at the same call site: the first N calls are given unique values, then the next call is given `Top`, but these can still all be distinguished. That might be relevant, for example, if you "know" that your program has 8 threads (or the autotuner says so): then N can be set to 7.

sim642 commented on src/util/gobList.ml:
```diff
@@ -30,6 +30,13 @@ let rec fold_while_some (f : 'a -> 'b -> 'a option) (acc: 'a) (xs: 'b list): 'a
 
 let equal = List.eq
 
+(** [span p xs] is [take_while p xs, drop_while p xs] but may be more efficient *)
+let span p =
+  let rec span_helper prefix = function
+    | x :: xs when p x -> span_helper (x :: prefix) xs
+    | suffix -> List.rev prefix, suffix
+  in span_helper []
```
This already exists as `BatList.span`.

sim642 commented on src/domains/printable.ml:
```diff
@@ -470,7 +470,7 @@ module type ChainParams = sig
   val names: int -> string
 end
 
-module Chain (P: ChainParams): S with type t = int =
```
Why is this signature removal required?

just-max replied:
Oops, it's not :) I had added signatures all over the place while tinkering, then removed them again, and I guess also accidentally removed this.

sim642 commented on src/cdomains/threadIdDomain.ml:
```diff
@@ -55,14 +69,16 @@ struct
 
   let is_main = function
-    | ({vname = "main"; _}, None) -> true
+    (* shouldn't this check configured mainfun?? *)
```
```suggestion
```
Since this check was changed.

sim642 commented on src/cdomains/threadIdDomain.ml:
```diff
@@ -30,21 +31,34 @@ sig
+      Printable.Option (
+        Printable.Prod
+          (Node) (
+          Printable.Option
+            (Printable.Int)
+            (struct let name = "no index" end)))
```
MallocWrapper used `Chain` for this, why doesn't that work here?

just-max replied:
The problem is that `Chain` is a functor, returning a different module for each `n`. For the query we need a single module, hence this `Int` lattice.

Now that you mention it though, it might be possible to put the thread wrapper chain in a seperate module and reference it in both the thread wrapper analysis and in queries, which would clean things up.

sim642 commented on src/analyses/threadId.ml:
```diff
@@ -87,13 +87,25 @@ struct
   let threadspawn ctx lval f args fctx =
     let (current, td) = ctx.local in
-    (current, Thread.threadspawn td ctx.prev_node f)
+    (* here we see the updated counter, so we want the previous counter value *)
+    let node, index = indexed_node_for_ctx ~previous:true ctx in
```
Just do the query in `fctx` instead of `ctx`. Then it's querying the counter from the created thread, not the creator thread.

sim642 commented on src/analyses/wrapperFunctionAnalysis.ml:
```diff
@@ -0,0 +1,230 @@
+(* The main analysis, generic to which functions are being wrapped. *)
+module SpecBase (UniqueCountArgs : UniqueCountArgs) (WrapperArgs : WrapperArgs) =
+struct
+  include Analyses.DefaultSpec
```
Might as well switch to `IdentitySpec` here to avoid all the identity transfer functions.

michael-schwarz commented:
@just-max Do you still plan to address the comments here, or should we do it ourselves?

just-max commented:
Thanks for the reminder on this ;) I've addressed all the comments (thanks Simmo!)

There's one last problem I can't figure out: if the `unique_thread_id_count` is too small (e.g. set to `0` when there are two threads), Goblint should not be able to distinguish threads created at the same location. After implementing Simmo's suggestion to use `fctx` though, this isn't the case. See the test `71/05`, which specifically checks that a race does occur if `unique_thread_id_count` is set to 0. This test passes before implementing the change, but fails after, and I can't figure out why Goblint considers the two threads in that test case to be distinct.

sim642 commented:
> See the test `71/05`, which specifically checks that a race does occur if `unique_thread_id_count` is set to 0. This test passes before implementing the change, but fails after, and I can't figure out why Goblint considers the two threads in that test case to be distinct.

When run with `--enable allglobs` to also see the non-racing accesses, it shows that the created threads are considered unique and thus joining "works". So Goblint actually only sees one unique thread being created (from the one location), not two. I didn't look into why that's the case though.

just-max commented:
In the wrapper function analysis' `threadenter`, the new thread receives a fresh view of the current wrapper node and counter. Then, when `threadspawn` goes to use `fctx` to add to the set of created threads, it adds the unwrapped node to the set of created threads.

https://github.com/goblint/analyzer/blob/68b22676ada38dda0aa61df7edc740118d7f283e/src/analyses/wrapperFunctionAnalysis.ml#L85-L87

https://github.com/goblint/analyzer/blob/68b22676ada38dda0aa61df7edc740118d7f283e/src/analyses/threadId.ml#L100-L103

I've changed `threadenter` in the wrapper function analysis to pass `ctx.local` instead of `bot`, but I'm not sure that's really the right approach.

sim642 commented:
> I've changed `threadenter` in the wrapper function analysis to pass `ctx.local` instead of `bot`, but I'm not sure that's really the right approach.

That doesn't seem right indeed because the idea has been that each thread is handled on its own, independent of who created it.

Now looking at it, why is `ThreadId.threadspawn` using `indexed_node_for_ctx` at all? `ThreadId.threadenter` already computed the created thread's correct ID before the wrapper analysis' `threadenter` sets it to `bot`. Couldn't `ThreadId.threadspawn` just use the already computed thread ID from `fctx.local`?

just-max commented:
The spawned thread's ID from `fctx.local` in `ThreadId.threadspawn` is a stateful thread ID (prefix-set if history is enabled), whereas the set of created threads contains just the thread creation nodes. So there's no way to directly add a node to the set given the abstract thread ID.

I've changed the `ThreadId` domain to also include the (wrapper) node that created the thread, so that it is available in `fctx` in addition to the abstract thread ID.

<details>
<summary>Alternative Idea</summary>

https://github.com/goblint/analyzer/compare/master...just-max:goblint-analyzer:thread_wrappers_stateful_id

Another option is: change `Stateful` so that it takes an abstract thread ID instead of a just a node. The prefix-set implementation uses the top element of the prefix stack for unique threads, and all the nodes in the suffix set for non-unique threads, to update the set of thread creation nodes. I'm not convinced this is a good idea though: for non-unique threads, this is quite an over-approximation.

</details>

just-max commented:
Conflicts are resolved. Thanks again for the help getting this done!
